### ISSUE TITLE
Psalm does not track aliasing

### ISSUE THREAD

**Issue by olleharstedt** [vimeo/psalm#3614]
Probably by design, because it's (very) hard to solve? Example of false negative (should fail): https://psalm.dev/r/8aeedae738

One possible solution could perhaps be to tag variables that should not be aliased, like `@psalm-no-alias`. Or that aliases are read-only and may not call non-pure methods of an object.

**Comment by muglug** [comment]
Yup, this is a _hard_ problem that cannot be solved in the general case in a performant manner.

**Comment by olleharstedt** [comment]
> Yup, this is a _hard_ problem that cannot be solved in the general case in a performant manner.

You guys don't happen to have a forum where ideas like this can be discussed in a more informal manner by contributors and users?

**Comment by muglug** [comment]
> You guys don't happen to have a forum where ideas like this can be discussed in a more informal manner by contributors and users?

Not yet – setting up discussions currently doesn't work.

I think you're asking for "Equality-based Pointer Analysis" – see [this paper](https://pure.au.dk/ws/files/82324148/feldthaus2014thesis.pdf) for a deeper dive. I don't think it would reveal too many new useful bugs, and it would _significantly_ add to Psalm's complexity.

**Comment by olleharstedt** [comment]
> I think you're asking for "Equality-based Pointer Analysis" – see [this paper](https://pure.au.dk/ws/files/82324148/feldthaus2014thesis.pdf) for a deeper dive. I don't think it would reveal too many new useful bugs, and it would increase Psalm's complexity quite a bit.

Thanks for the link! And yes, tracking aliases only makes sense together with typestate (@self-out etc etc). Don't know if your plugin system allows such extensions, will read up on it. :)

**Comment by olleharstedt** [comment]
Another alternative is ownership inference, or manual ownership (Rust). There are some papers about this online, e.g. with so called Universe Types.

**Comment by olleharstedt** [comment]
One problem to use escape analysis to track ownership is whole-program analysis vs modular analysis. How can a function or method know its access to a variable without first analyzing its context?

**Comment by olleharstedt** [comment]
Another alternative: `@psalm-borrower-readonly`, which makes a class modifiable only for the owner. How to track ownership? It can be made fairly simple without move semantics (not possible to transfer ownership). Then this would create ownership:

```php
$file = new File()
```

and this would create a borrowing variable:

```php
$file2 = $file
```

Borrowing is the default for all types that are tagged with `@psalm-borrower-readonly`. All function arguments are always borrowing.

Things that would not be possible:

* Factory pattern

But maybe some simple statements could be allowed to transfer owner ship? Like `return new File();`. But that would require annotation for ownership, so that the function signature is aware.
